### PR TITLE
[extension/jaegerremotesampling] Support reload_interval option in remote mode of usage

### PR DESCRIPTION
**Pull request by zcross**
**Description:** <Describe what has changed.>
Updates the jaegerremotesampling extension's `remote` mode of usage to support the `reload_interval` caching option already supported in `file` mode of usage.

**Link to tracking Issue:** #24840

**Testing:** 
* Integration test updates show collector<>remote interaction with and without caching
* Unit tests cover specific caching behavior inclusive of concurrency (leak detector / race detector etc)

**Documentation:** Existing readme snippet referring to the `reload_interval` config option explains that it works for both modes of `remote`.

(Reopened after closing #24854 – after addressing its incremental feedback – to double check my EasyCLA settings)

### CODE REVIEW

File: extension/jaegerremotesampling/internal/remote_strategy_cache.go
Comment (MovieStoreGuy) [inline extension/jaegerremotesampling/internal/remote_strategy_cache.go:33]: More of a nit, what is the mutex protecting? Also, I would suggest naming it `rw` instead of `mu` since that is the default for `RWMutex`.

More of a stylistic thing is ordering the fields so that mutex sits as a hat on top of fields it protects, like:

```golang
type serviceStrategyTTLCache struct {
  itemTTL time.Duration
  clockFn func()time.Time

  rw sync.RWMutext
  stopChn chan struct{}
  items map[string]...
}
```

File: extension/jaegerremotesampling/internal/remote_strategy_cache.go
Comment (MovieStoreGuy) [inline extension/jaegerremotesampling/internal/remote_strategy_cache.go:61]: I suggest using https://github.com/tilinna/clock since you can reference the context to see if there is a mock clock defined. Saves having to make sure a time function is set.

So the function would look like:

```golang
func (c *serviceStratergyTTLCache) get(ctx context.Context, serviceName string) (*sampling.SamplingStrategyResponse, bool) {
  c.mu.RLock()
  defer c.mu.RUnlock()

  found, ok := c.Items[serviceName]
  if ok && clock.Now(ctx).Before(found.retrivedAt.Add(c.ItemTTL)) {
     return found.stratergyResponse, true
  }
  return nil, false
}
```

File: extension/jaegerremotesampling/internal/remote_strategy_store.go
Comment (MovieStoreGuy) [inline extension/jaegerremotesampling/internal/remote_strategy_store.go:54]: An idiomatic go error keeps the error message text as lower case to ensure that when it is logged like
`fmt.Println("An error happened:", err)` that it reads like a normal sentence.

I suggest renaming to something like `remote call failed: %w`

**Review by MovieStoreGuy — COMMENTED** [review]
The changes seem reasonable to me, I only have stylistic comments to recommend here.

Otherwise, if the changes are good with the code owners then I see no problem merging.

**Comment by zcross** [comment]
Rebased/squashed `main` after addressing latest round of feedback

**Comment by zcross** [comment]
Build should be green, the single failing lint test case just timed out.

**Comment by zcross** [comment]
ICYMI @frzifus , I closed the first instance of this PR and reopened this new one (I wanted to double check my EasyCLA, and it looks good).

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by zcross** [comment]
@frzifus if you land your thrift PR and I rebase/clean this up, do you think it's feasible to get this _and_ yours into `0.83.0`? Of course, I'll wait

**Comment by zcross** [comment]
Thanks for the review @MovieStoreGuy

**Review by frzifus — APPROVED** [review]

**Comment by zcross** [comment]
@frzifus I just saw your approval above. I would be ecstatic to get this onto `main` ahead of the `0.83.0` cut, but I'm wondering: do you still want me to wait until #18485 is merged to merge this? I don't want to complicate your work on that because this PR is an optimization more than a "must have" for me (which the HTTP header passing PR was).

**Comment by frzifus** [comment]
No please go ahead. If we follow the copy approach its easy to rebase :)

**Comment by zcross** [comment]
> No please go ahead. If we follow the copy approach its easy to rebase :)

Awesome!! thank you

File: extension/jaegerremotesampling/internal/remote_strategy_cache.go
Comment (atoulme) [inline extension/jaegerremotesampling/internal/remote_strategy_cache.go:67]: do you need to defer? Can you let go of the lock a bit earlier here?

Comment (zcross) [reply extension/jaegerremotesampling/internal/remote_strategy_cache.go:67]: don't need to, it's just a pattern I've seen (Go newbie here). In this case the subsequent branches are almost all immediate returns based on some in-memory state so I don't think lock contention is that big of a deal, so I lean toward the code cleanliness / lazy approach of treating the entire function as synchronized rather than having some fine grained locking around critical sections inside of it.

File: extension/jaegerremotesampling/internal/remote_strategy_cache.go
Comment (atoulme) [inline extension/jaegerremotesampling/internal/remote_strategy_cache.go:54]: so here you're saying that we will not clean anything for a minimum of 5 minutes, no matter what TTL we have set.
Can we make this configurable instead?

Comment (atoulme) [reply extension/jaegerremotesampling/internal/remote_strategy_cache.go:54]: I guess I'm more taking issue with the comments line 92 to 96 that states that the cleanup interval is proportional to TTL, when it's not if TTL is less than 5 minutes.

It also means that there is a window of 5 minutes where the buffer can grow unbounded and no one can do much about it.

You start this cache with 32 entries - what's the expected type of volume you'd see go through that cache?

Comment (zcross) [reply extension/jaegerremotesampling/internal/remote_strategy_cache.go:54]: The cache is keyed off of service names; in practice I would expect the number of items to be low (hundreds or thousands of values). This approach is defending against rarely queried items remaining in the cache and never getting expired out (replaced) if they are not queried. About volume of requests per collector, I expect it to be a function of the number of instrumented services, their respective agent configuration that controls agent polling, and the number of collectors deployed in the environment. I think I'll tweak it to just run on the configured TTL to simplify this in light of this conversation.

Comment (zcross) [reply extension/jaegerremotesampling/internal/remote_strategy_cache.go]: How's what I just pushed?

It:
* Runs the cleaner at a frequency equal to the per item cache TTL
* Prunes stale items
* Recreates the map to prevent strictly increasing map growth (as unlikely as that may be given the expectations about service name cardinality in a typical deployment)

Hoping it's a bit more intuitive while still fulfilling my intent about managing garbage associated w/ the cache.

Comment (zcross) [reply extension/jaegerremotesampling/internal/remote_strategy_cache.go]: > It also means that there is a window of 5 minutes where the buffer can grow unbounded and no one can do much about it.

BTW: Regarding this part of the concerns, the default behavior is no caching (it's inert/off by default) and somebody seeing problems created by caching can just turn it off via a extension config change.

And in case it was overemphasized, this thing growing unbounded would mean receiving a steady stream of cache-missing queries meaning you have an extremely large set of service _names_ (not just pods/host, but service identities) – a scenario I consider very unlikely but still defending against by this change.

Comment (zcross) [reply extension/jaegerremotesampling/internal/remote_strategy_cache.go]: I'm gonna remove the comment that even mentions unbounded growth because it sounds a lot worse than it is.

**Review by atoulme — APPROVED** [review]
LGTM

**Pull request merged**
